File: Data Structures & Algorithms/max-water-container/submission-1.cpp
class Solution {
   public:
    int maxArea(vector<int>& heights) {
        int const size = static_cast<int>(heights.size());
        int result = 0;
        int left = 0;
        int right = size - 1;

        while (left < right) {
            int area = 0;

            if (heights[left] < heights[right]) {
                area = (right - left) * heights[left];
                left++;
            } else if (heights[left] > heights[right]) {
                area = (right - left) * heights[right];
                right--;
            } else {
                area = (right - left) * heights[left];
                left++;
                right--;
            }

            result = max(result, area);
        }

        return result;
    }
};

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        result = 0
        stack = []

        for i, h in enumerate(heights + [-1]):
            while stack and heights[stack[-1]] > h:
                height = heights[stack.pop()]
                if not stack:
                    left = 0
                else:
                    left = stack[-1] + 1
                right = i
                result = max(result, (right - left) * height)

            stack.append(i)

        return result
